### PR TITLE
GEODE-5594: Enable endpoint identification during using SSL handshake.

### PR DESCRIPTION
**Pull request by sboorlagadda**
User can enable endpoint identification during SSL handshake
   to validate server's hostname with server's identity.

   New boolean property ssl-endpoint-validation-enabled is added.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/distributed/ConfigurationProperties.java [hunk 1]
```diff
@@ -1929,6 +1929,17 @@
...
+   * The static String definition of the <i>"ssl-endpoint-identification-enabled"</i> property <a
+   * name="ssl-endpoint-identification-enabled"/a>
+   * </p>
+   * <U>Description</U>: If true, the endpoint identification will be enabled during SSL/TLS
+   * handshake.
+   * </p>
+   * <U>Default</U>: code>"false"</code>
+   * <U>Since</U>: Geode 1.8
+   */
+  String SSL_ENDPOINT_IDENTIFICATION_ENABLED = "ssl-enable-endpoint-identification";
```
Comment (upthewaterspout) [inline geode-core/src/main/java/org/apache/geode/distributed/ConfigurationProperties.java]: Can you add some more details to the description of this? Like how it interacts with truststores and keystores, which which connections this affects and why you would want to turn it on?

**Comment by mcmellawatt** [comment]
This may have already been caught, but the PR title is "GEDOE" instead of "GEODE"

File: geode-core/src/main/java/org/apache/geode/distributed/ConfigurationProperties.java (hunk 1 above)
Comment (sboorlagadda) [reply geode-core/src/main/java/org/apache/geode/distributed/ConfigurationProperties.java]: I have added details about hostname verification. This does not change the current interactions with truststore and keystore.

**Comment by sboorlagadda** [comment]
@bschuchardt @upthewaterspout Thanks for your review. I have added few negative test cases to show that when `ssl-endpoint-identification-enabled=true` then TLS clients validate hostname and fails to connect if hostname is not available in server's certificate.

**Comment by sboorlagadda** [comment]
@bschuchardt 

Added a low-level test(TestSSLClientIntegration.java) to specifically test SSL handshake to validate hostname validation and assert exception reason. Also, the existing WAN test is testing the positive case and made it more explicit about enabling the flag. This new low-level test should cover all cases for any high-level uses of TcpClient (client-server or wan or Gfsh etc).

**Comment by sboorlagadda** [comment]
@mcmellawatt Thanks, Ryan. Corrected the title.

**Pull request merged**